src/test/java/com/aws/greengrass/clientdevices/auth/certificate/BackgroundCertificateRefreshTest.java

```diff
@@ -190,4 +215,79 @@ void GIVEN_certsAndThings_WHEN_certificateAssociatedToMoreThanOneThing_THEN_itDo
         assertEquals(pemStore.getPem(certA.getCertificateId()).get(), certificateAPem);
     }
 
+    @Test
+    void GIVEN_networkWasDown_WHEN_networkUp_THEN_backgroundTaskTriggered() {
+        backgroundRefresh.accept(NetworkState.ConnectionState.NETWORK_DOWN);
+        verify(backgroundRefresh, times(0)).run();
+        backgroundRefresh.accept(NetworkState.ConnectionState.NETWORK_UP);
+        verify(backgroundRefresh, times(1)).run();
+        backgroundRefresh.accept(NetworkState.ConnectionState.NETWORK_DOWN);
+        verify(backgroundRefresh, times(1)).run();
+    }
+
+    @Test
+    void GIVEN_networkChangesAndScheduledByTime_WHEN_triggered_THEN_shouldNotRunMoreThanOnceADay() {
+        Instant now = Instant.now();
+        mockInstant(now.toEpochMilli());
+
+        verify(iotAuthClientFake, times(0)).getThingsAssociatedWithCoreDevice();
+        backgroundRefresh.run();
+        backgroundRefresh.run();
+        backgroundRefresh.run();
+        verify(iotAuthClientFake, times(1)).getThingsAssociatedWithCoreDevice();
+
+        Instant twentyFiveHoursLater = Instant.now().plus(Duration.ofHours(25));
+        mockInstant(twentyFiveHoursLater.toEpochMilli());
+
+        backgroundRefresh.run();
+        verify(iotAuthClientFake, times(2)).getThingsAssociatedWithCoreDevice();
+    }
+
+    @Test
+    void GIVEN_triggeredByMultipleThreads_WHEN_called_THEN_itShouldOnlyRunOnceForThatTimeWindow()
+            throws InterruptedException {
+       Thread t1 = new Thread(() -> { backgroundRefresh.run(); });
+       Thread t2 = new Thread(() -> { backgroundRefresh.run(); });
+       t1.start();
+       t2.start();
+       t1.join();
+       t2.join();
+
+        verify(iotAuthClientFake, times(1)).getThingsAssociatedWithCoreDevice();
+    }
+
+    @Test
+    void GIVEN_triggeredInDifferentInstants_WHEN_triggeredExactlyOneDayLater_THEN_itShouldRunAgain() {
+        Instant now = Instant.now();
+        mockInstant(now.toEpochMilli());
+
+        backgroundRefresh.run();
+
+        Instant aDayLater = now.plus(Duration.ofHours(24));
+        mockInstant(aDayLater.toEpochMilli());
+
+        backgroundRefresh.run();
+        verify(iotAuthClientFake, times(2)).getThingsAssociatedWithCoreDevice();
+    }
+
```

Comment:
I think there's a missing test case here. If the network call fails, it should be retried on the next network up event

Reply:
K this should be testing that scenario now https://github.com/aws-greengrass/aws-greengrass-client-device-auth/pull/175/files#diff-c6f02a506e0f67706541eefa03ba3a070eb33963da172e13ee67b590423d1f83R232

Reply:
Basically if it fails it can retry the net time it gets called (which could be triggered by an network up or the scheduler)

Reply:
This gets tested now here https://github.com/aws-greengrass/aws-greengrass-client-device-auth/pull/175/files#diff-c6f02a506e0f67706541eefa03ba3a070eb33963da172e13ee67b590423d1f83R231. I did not use a InternalServer exception because then it enters the retry policy and hangs in there forever